anvio/scgsdatabase.py

```diff
@@ -0,0 +1,791 @@
+
+
+#!/usr/bin/env python3
+
+import os
+import gzip
+import shutil
+import requests
+from io import BytesIO
+import anvio
+import pickle
+import re
+import shutil
+import subprocess
+import sys
+import tarfile
+import anvio.fastalib as u
+from anvio.drivers.diamond import Diamond
+import anvio.terminal as terminal
+import anvio.pfam as pfam
+import anvio.dbops as dbops
+import anvio.utils as utils
+import anvio.terminal as terminal
+import anvio.filesnpaths as filesnpaths
+
+from anvio.errors import ConfigError
+
+
+run = terminal.Run()
+progress = terminal.Progress()
+pp = terminal.pretty_print
+run_quiet = terminal.Run(log_file_path=None,verbose=False)
+progress_quiet = terminal.Progress(verbose=False)
+
+
+#run_quiet = terminal.Run(verbose=False)
+#progress_quiet = terminal.Progress(verbose=False)
+
+_author__ = "Developers of anvi'o (see AUTHORS.txt)"
+__copyright__ = "Copyleft 2015-2018, the Meren Lab (http://merenlab.org/)"
+__license__ = "GPL 3.0"
+__version__ = anvio.__version__
+__maintainer__ = "Quentin Clayssen"
+__email__ = "quentin.clayssen@gmail.com"
+
+
+run = terminal.Run()
+progress = terminal.Progress()
+pp = terminal.pretty_print
+run_quiet = terminal.Run(verbose=False)
+progress_quiet = terminal.Progress(verbose=False)
+
+
+### open fasta once.... make db of Bacteria_71 fasta .......
+class scgsdatabase():
+    def __init__(self, args, run=run, progress=progress):
+        self.args = args
+        self.run = run
+        self.progress = progress
+
+        A = lambda x: args.__dict__[x] if x in args.__dict__ else None
+
+        self.genes_files_directory = args.genes_files_directory
+
+        self.scgs_directory = args.scgs_directory
+
+        #self.num_threads=args.num_threads
+
+
+
+        self.dictionnary_level=self.dictionnary_level()
+
+
+
+        if not args.taxofiles:
+            self.taxofiles = os.path.join(os.path.dirname(anvio.__file__), 'data/misc/SCG/merge_taxonomy.tsv')
+
+        #if not args.outtsv:
+        self.outtsv = os.path.join(os.path.dirname(anvio.__file__), 'data/misc/SCG/mergedb/matching_taxonomy.tsv')
+
+        #if not args.num_threads:
+        self.num_threads=args.num_threads
+
+        if not args.num_threads:
+            self.num_threads=2
+
+
+        self.hmms=args.hmms
+
+        if not args.genes_files_directory:
+            self.genes_files_directory = os.path.join(os.path.dirname(anvio.__file__), 'data/misc/SCG/msa_individual_genes')
+
+        self.genesfiles=os.listdir(self.genes_files_directory)
+
+
+        if not args.outputdirectory:
+            outputdirectory = os.path.join(os.path.dirname(anvio.__file__), 'data/misc/SCG/mergedb')
+            if not os.path.exists(outputdirectory):
+                os.mkdir(os.path.dirname(anvio.__file__)+'/data/misc/SCG/mergedb')
+            self.outputdirectory = outputdirectory
+
```

Comment:
in the code above you don't check whether the user have any rights to open a directory in that path. these will ALL EXPLODE when run on many computers. there is not a single exception handling throughout the entire file, and this is a very serious concern to me. 

it is easy to write code that runs on your computer. it is harder to write code that is elegant and still runs. it is much harder to make that code be able to run ALSO on others' computers. it is much much much harder to write code that knows why it fails when it fails, so users don't need a developer to decrypt some awkward message when the code fails and they can figure out from the exception they get. anvi'o codebase strives to reach to that ideal. your module is currently at the first stage, and we will have to make sure it is comparable to the rest of the codebase before it is merged to `master`.

how are we going to get there? it is two folds: (1) don't reinvent the wheel and work with the rest of the codebase, and (2) anticipate points of failure and add try/except procedures to catch them. 

For the first pint, anvi'o has functions for most of these kinds of tasks. for instance, here is one that does what you are trying to do in the lines above:

https://github.com/merenlab/anvio/blob/master/anvio/filesnpaths.py#L234

Whenever you are trying to do something this trivial, you should search the codebase assuming that someone already has done it, ask Ozcan or me if you can't find something, and then consider implementing that atomic trivial task (such as creating an output directory for something) as a function in a central module so others can use later.

Second, infest the entire codebase with try/except procedures and try to inherit an exception-driven mindset. For instance, open this module and search for the keyword `raise` and compare it to your module:

https://github.com/merenlab/anvio/blob/master/anvio/filesnpaths.py

There are 1219 raise statements in the codebase. You certainly need much more of them here :)